File: math/9421/main.cpp
// Authored by : ccocco0609
// Co-authored by : -
// Link : http://boj.kr/f9b92c1f899a4146bdf39d4e2a4a9f77

#include <bits/stdc++.h>
using namespace std;

bool prime(int n) {

	if (n < 2) {
		return false;
	}
	else {
		for (int i = 2; i * i <= n; i++ ) {
			if (n % i == 0) {
				return false;
			}
		}
		return true;
	}

}

void ans(int n) {
	int tmp = n;
	int next = 0;
	map<int, int> record;
	while (true) {
		while(tmp) {
			next = next + pow(tmp % 10, 2);
			tmp = tmp / 10;
		}
		tmp = next;
		next = 0;
		if (tmp == 1) {
			cout << n << '\n';
			return;
		}
		if (record.find(tmp) != record.end()) {
			return;
		}
		record.insert(make_pair(tmp, 0));
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	int n = 0;

	cin >> n;

	for (int i = 2; i <= n; i++) {
		if (prime(i)) {
			ans(i);
		}
	}


	return 0;
}

File: math/9421/Main.java
//  Authored by : vswngjs
//  Co-authored by : -
//  Link : http://boj.kr/b9801e796f7547e890b0eb1231a9bce2
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String [] args) {
        FastReader rd = new FastReader();
        int value = rd.nextInt();
        boolean prime[] = new boolean[1000001];
        for (int i = 2; i <= 1000000; i++) {
            for (int j = i * 2; j <= 1000000; j +=i) {
                prime[j] = true;
            }
        }
        for (int i = 3; i <= value; i++) {
            if(!prime[i]) {
                Set<Integer> set = new HashSet<>();
                String str = Integer.toString(i);
                boolean flag = false;
                while(true) {
                    int sum = 0;
                    for (int j = 0; j < str.length(); j++) {
                        sum += Math.pow(str.charAt(j) -'0', 2);
                    }
                    if(sum == 1) {
                        flag = true;
                        break;
                    }
                    if(!set.contains(sum)){
                        set.add(sum);
                        str = String.valueOf(sum);
                    } else break;
                }
                if(flag) System.out.println(i);
            }
        }

    }
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
